src/device/router_device_routing.erl

```diff
@@ -944,6 +967,61 @@ find_device(PubKeyBin, DevAddr, MIC, Payload, Chain) ->
             {Device, NwkSKey}
     end.
 
+%%%-------------------------------------------------------------------
+%% @doc
+%% If there are devaddr collisions, we don't know the correct device until we
+%% get the whole packet with the payload. Once we have that, the device can tell
+%% routing the a packet hash belongs to it. That reservation will only last a
+%% short while. But it might be enough to have subsequent offers lookup the
+%% correct device by the time they come through.
+%% @end
+%%%-------------------------------------------------------------------
+-spec cache_device_for_hash(PHash :: binary(), Device :: router_device:device()) -> ok.
+cache_device_for_hash(PHash, Device) ->
+    ok = e2qc_nif:put(
+        ?PHASH_TO_DEVICE_CACHE,
+        PHash,
+        erlang:term_to_binary(Device),
```

Comment:
we have a function for this `router_device:serialize/1`

Reply:
updated, and also using `router_devcie:deserialize/1` in the matching `get`